homeassistant/components/stream/core.py

```diff
@@ -3,45 +3,102 @@
 
 import asyncio
 from collections import deque
-import io
-from typing import TYPE_CHECKING, Any, Callable
+import datetime
+from typing import Any, Callable
 
 from aiohttp import web
+import async_timeout
 import attr
 
 from homeassistant.components.http import HomeAssistantView
 from homeassistant.core import HomeAssistant, callback
 from homeassistant.helpers.event import async_call_later
 from homeassistant.util.decorator import Registry
 
-from .const import ATTR_STREAMS, DOMAIN
-
-if TYPE_CHECKING:
-    import av.container
-    import av.video
+from .const import ATTR_STREAMS, DOMAIN, MIN_SEGMENT_DURATION
 
 PROVIDERS = Registry()
 
 
-@attr.s
-class StreamBuffer:
-    """Represent a segment."""
+@attr.s(slots=True)
+class Part:
+    """Represent a segment part."""
 
-    segment: io.BytesIO = attr.ib()
-    output: av.container.OutputContainer = attr.ib()
-    vstream: av.video.VideoStream = attr.ib()
-    astream = attr.ib(default=None)  # type=Optional[av.audio.AudioStream]
+    duration: float = attr.ib()
+    independent: bool = attr.ib()
+    # http_start is offset from the first_moof_loc (not the start of the mp4 file)
+    # this is essentially in "http range" relative terms
+    http_range_start: int = attr.ib()
+    data: bytes = attr.ib()
 
 
-@attr.s
+@attr.s(slots=True)
 class Segment:
     """Represent a segment."""
 
-    sequence: int = attr.ib()
-    segment: io.BytesIO = attr.ib()
-    duration: float = attr.ib()
+    sequence: int = attr.ib(default=0)
+    # Since we need concurrent access to the segment in hls and worker, we copy out
+    # parts of the segment as bytes in init and in parts. We end up discarding segment.
+    init: bytes = attr.ib(default=None)
+    duration: float = attr.ib(default=0)
     # For detecting discontinuities across stream restarts
     stream_id: int = attr.ib(default=0)
+    parts: list[Part] = attr.ib(factory=list)
+    # Mapping of byte start to part index for parts
+    byterange_part_map: dict[int, int] = attr.ib(factory=dict)
+    # Where the first moof starts
+    first_moof_loc: int = attr.ib(default=0)
+    # This value is the current length of the segment
+    last_write_pos: int = attr.ib(default=0)
+    start_time: datetime.datetime = attr.ib(factory=datetime.datetime.utcnow)
+
+    @property
+    def complete(self) -> bool:
+        """Return whether the Segment is complete."""
+        return self.duration > 0
+
+    def get_bytes(self) -> bytes:
+        """Return reconstructed data for entire segment as bytes."""
+        return self.init + b"".join([part.data for part in self.parts])
+
+    def get_http_range_bytes(self, http_start: int, http_stop: int) -> bytes:
+        """Return bytes corresponding to the given http byte range."""
```

Comment:
Is this function effectively a more efficient version of something like `get_bytes()[http_start:http_stop]`?

Reply:
Not quite, it's also offset by `first_moof_loc` since we don't want to include the init section, so something similar to `get_bytes()[http_start+first_moof_loc,http_stop+first_moof_loc]`

Reply:
Could rename it to `get_bytes_without_init` as that might be more clear. There is also the caveat that the ranges endpoints should correspond to part boundaries.
Can probably refactor this to just build up the list instead of getting the start and end indexes.

Reply:
I'm going to reread the spec again and review again.

Reply:
This doesn't really have to do with the spec per se - it's more about fmp4 fragments and also about the new CMAF chunks. Just for further info:
- get_bytes() is used when we need to get the whole Segment (including the init) as an mp4 file. We only need this when we need to reconstruct the entire mp4 again (We currently only use this when feeding the file back in to pyav in `recorder.py`).
- The existing HLS spec supports TS or fragmented mp4 format. We previously converted stream to use fmp4 fragments. With fmp4, we serve the init section separately from the rest of the fragment. The fmp4 fragment itself may consist of many CMAF chunks, but in HLS the entire fragment is served as one piece.
- The LL-HLS spec supports serving the fmp4 fragment as separate chunks. We can do this by either serving them using different urls (named with a different part sequence per chunk) or by serving them via byteranges on the original HLS url.

So it makes sense for get_bytes() to be different from the other accesses because since all of the other accesses don't need init, they will be offset by `first_moof_loc`.

These functions could just stay inline in the hls code since there is minimal code reuse (they are mostly each just used once) and this would avoid the overhead of the function call, but putting them in one place does make it easier to maintain and maybe easier to read.